db/src/db.rs

```diff
@@ -285,60 +283,62 @@ pub fn ensure_current_version(conn: &mut rusqlite::Connection) -> Result<i32> {
+            (5, &rusqlite::types::Value::Real(ref x)) => Ok(TypedValue::Double((*x).into())),
+            (10, &rusqlite::types::Value::Text(ref x)) => Ok(TypedValue::String(x.clone())),
+            (13, &rusqlite::types::Value::Text(ref x)) => Ok(TypedValue::Keyword(x.clone())),
+            (_, value) => bail!(ErrorKind::BadSQLValuePair(value.clone(), *value_type_tag)),
+        }
+    }
 
-/// A `Value` that can be converted `ToSql`.
-///
-/// This is just working around Rust's refusal to allow us to implement a trait we don't originate
-/// for a type we don't originate.
-///
-/// TODO: &Value for efficiency.
-struct SqlValueWrapper(Value);
-
-impl ToSql for SqlValueWrapper {
-    fn to_sql(&self) -> rusqlite::Result<ToSqlOutput> {
-        match self.0 {
-            Value::Nil => Ok(ToSqlOutput::from(rusqlite::types::Value::Null)),
-            Value::Boolean(x) => Ok(ToSqlOutput::from(x)),
-            Value::Integer(x) => Ok(ToSqlOutput::from(x)),
-            // TODO: consider using a larger radix to save characters.
-            Value::BigInteger(ref x) => Ok(ToSqlOutput::from(x.to_str_radix(10))),
-            Value::Float(ref x) => Ok(ToSqlOutput::from(x.into_inner())),
-            // TODO: try to avoid this clone.
-            Value::Text(ref x) => Ok(ToSqlOutput::from(x.clone())),
-            Value::PlainSymbol(ref x) => Ok(ToSqlOutput::from(x.to_string())),
-            Value::NamespacedSymbol(ref x) => Ok(ToSqlOutput::from(x.to_string())),
-            Value::Keyword(ref x) => Ok(ToSqlOutput::from(x.to_string())),
-            Value::NamespacedKeyword(ref x) => Ok(ToSqlOutput::from(x.to_string())),
-            Value::Vector(_) => Err(rusqlite::Error::InvalidColumnName(format!("Cannot convert to_sql: {:?}", self.0))),
-            Value::List(_) => Err(rusqlite::Error::InvalidColumnName(format!("Cannot convert to_sql: {:?}", self.0))),
-            Value::Set(_) => Err(rusqlite::Error::InvalidColumnName(format!("Cannot convert to_sql: {:?}", self.0))),
-            Value::Map(_) => Err(rusqlite::Error::InvalidColumnName(format!("Cannot convert to_sql: {:?}", self.0))),
+    /// Given an EDN `value`, return a corresponding Mentat `TypedValue`.
+    ///
+    /// An EDN `Value` does not encode a unique Mentat `ValueType`, so the composition
+    /// `from_edn_value(first(to_edn_value_pair(...)))` loses information.  Additionally, there are
+    /// EDN values which are not Mentat typed values.
+    ///
+    /// This function is deterministic.
+    pub fn from_edn_value(value: &Value) -> Option<TypedValue> {
+        match value {
+            &Value::Boolean(x) => Some(TypedValue::Boolean(x)),
+            &Value::Integer(x) => Some(TypedValue::Long(x)),
+            &Value::Float(ref x) => Some(TypedValue::Double(x.clone())),
+            &Value::Text(ref x) => Some(TypedValue::String(x.clone())),
+            &Value::NamespacedKeyword(ref x) => Some(TypedValue::Keyword(x.to_string())),
```

Comment:
I think `TypedValue::Keyword` should contain a `Keyword`, preserving the namespace/name distinction and avoiding consing up a new string. This might mean you need to extract a `keyword` crate out of `edn` for reuse. Please file an issue.